### PR TITLE
Add local follows | Convert some Kraken functions to Helix

### PR DESCRIPTION
**Pull request by Nickwasused**
# How this works
With Account:
- Your Follows get imported from Twitch
- You can´t unfollow someone that you followed on Twitch
- Wiping follows dosen´t remove Twitch follows from the Database
- You can follow new Channels from within the App
- You can unfollow the Channels that you followed in the App
- Follows from within the App are not synced (because #230)

Without Account:
- You can do nothing, because of Helix.
~- You can follow new Channels from within the App~
~- You can unfollow the Channels that you followed in the App~
~- Your Channels get shown in the My Streams Tab when they are live~

# Problems with Helix
- The Viewer Count of "Top Games" cannot be fetched (https://twitch.uservoice.com/forums/310213-developers/suggestions/37393114-add-viewer-and-channel-count-in-helix-games-top).
- The Vod Chat cannot be fetched (https://twitch.uservoice.com/forums/310213-developers/suggestions/44063160-vod-chat-for-helix)

# Showcase

## With Account
[youtube](https://www.youtube.com/watch?v=MXUcvombeBc)
- Ubisoft and Bobross are Channels that got imported from Twitch

## Without Account
[youtube](https://www.youtube.com/watch?v=0gz5XRPARWA)
- Bob Ross is used as Showcase Channel

# Problems
As mentioned in #236 when V5 is shutting down then everything is going to require an Auth Token. So you can´t make the required API requests without a Twitch Account logged in. But local follows should still be working with the new API.

# Fixing
This should be closing:
#239 
#230
(Maybee #81 and #124)

# To-Do
- [x] ~wait for helix rewrite/support (#236)~ Done in 3a1684851a9ccf94757efdb861886fb0d354fe9e
- [x] Fix "GetTwitchUserFollows"
- [x] Fix importing Channels from Twitch

### CODE REVIEW

**Comment by RickyM7** [comment]
This is amazing. A question, will an option be added to export/import the list of followed channels?

**Comment by Nickwasused** [comment]
I have added basic export/import functions in https://github.com/twireapp/Twire/pull/240/commits/acc3ac7451ddbc23403481bcdfc0e72e3be126a8

The follows get exported to the Twire data folder.

**Comment by RickyM7** [comment]
> I have added basic export/import functions in acc3ac7
> 
> The follows get exported/imported to/from the Twire data folder.

I just tested it. Export works normally, but import doesn't work and the app even crashes sometimes when using this option.

**Comment by RickyM7** [comment]
The error continues (I tested the last apk generated). I also have a suggestion, is there a way to make a popup message appear whenever we export/import channels?

``
time: 1628769584734
msg: java.lang.AssertionError: Could not execute method for android:onClick
stacktrace: java.lang.IllegalStateException: Could not execute method for android:onClick
	at android.view.View$DeclaredOnClickListener.onClick(View.java:6268)
	at android.view.View.performClick(View.java:7448)
	at com.balysv.materialripple.MaterialRippleLayout$PerformClickEvent.run(MaterialRippleLayout.java:648)
	at com.balysv.materialripple.MaterialRippleLayout.onTouchEvent(MaterialRippleLayout.java:227)
	at android.view.View.dispatchTouchEvent(View.java:14309)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3112)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2785)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:488)
	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1873)
	at android.app.Activity.dispatchTouchEvent(Activity.java:4125)
	at androidx.appcompat.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:69)
	at androidx.appcompat.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:69)
	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:446)
	at android.view.View.dispatchPointerEvent(View.java:14568)
	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:6027)
	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:5825)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5316)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5373)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5339)
	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5491)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5347)
	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:5548)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5320)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5373)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5339)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5347)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5320)
	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:8091)
	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:8042)
	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:8003)
	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:8214)
	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:220)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:335)
	at android.os.Looper.loop(Looper.java:183)
	at android.app.ActivityThread.main(ActivityThread.java:7665)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:594)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at android.view.View$DeclaredOnClickListener.onClick(View.java:6263)
	... 54 more
Caused by: java.lang.AssertionError
	at com.perflyst.twire.service.SubscriptionsDbHelper.onImport(SubscriptionsDbHelper.java:116)
	at com.perflyst.twire.activities.settings.SettingsGeneralActivity.onImport(SettingsGeneralActivity.java:168)
	... 56 more
``

**Comment by RickyM7** [comment]
Tested f2f24d8 artifact. The Import option still doesn't work.

``
time: 1628787268123
msg: java.lang.AssertionError
stacktrace: java.lang.AssertionError
	at com.perflyst.twire.service.SubscriptionsDbHelper.onImport(SubscriptionsDbHelper.java:134)
	at com.perflyst.twire.activities.settings.SettingsGeneralActivity.lambda$onImport$7$SettingsGeneralActivity(SettingsGeneralActivity.java:191)
	at com.perflyst.twire.activities.settings.-$$Lambda$SettingsGeneralActivity$8UY9iNiS7RYCzRPXobFxZ1XwcPs.onClick(Unknown Source:2)
	at com.afollestad.materialdialogs.MaterialDialog.onClick(MaterialDialog.java:343)
	at android.view.View.performClick(View.java:7448)
	at android.view.View.performClickInternal(View.java:7425)
	at android.view.View.access$3600(View.java:810)
	at android.view.View$PerformClick.run(View.java:28305)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7665)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:594)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
``

File: app/src/main/java/com/perflyst/twire/tasks/GetStreamsCountTask.java [hunk 1]
```diff
@@ -23,10 +24,22 @@ public GetStreamsCountTask(Context context, Delegate delegate) {
     @Override
     protected Integer doInBackground(Void... params) {
         try {
-            final String URL = "https://api.twitch.tv/kraken/streams/followed?oauth_token=" + settings.getGeneralTwitchAccessToken() + "&offset=0&stream_type=live";
+            // build the api link
+            String kraken_url = "https://api.twitch.tv/kraken/streams?limit=100&stream_type=live&offset=0&channel=";
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/tasks/GetStreamsCountTask.java:28]: This should probably be moved to the [new Helix API](https://dev.twitch.tv/docs/api/reference#get-streams) and add support for pagination so the count doesn't stop at 100.

File: app/src/main/java/com/perflyst/twire/misc/FollowHandler.java [hunk 2]
```diff
@@ -17,64 +14,53 @@
...
+    // isStreamerFollowed, isStreamerTwitch need to be checked every time, because they can change fast for example when spamming the button
     public boolean isStreamerFollowed() {
+        boolean isUserFollowed = Service.isUserFollowingStreamer(mChannelInfo.getStreamerName(), mContext);
+        isStreamerFollowed = isUserFollowed;
         return isStreamerFollowed;
     }
+    public boolean isStreamerTwitch() {
+        boolean isUserTwitch = Service.isUserTwitch(mChannelInfo.getUserId(), mContext);
+        isStreamerTwitch = isUserTwitch;
+        return isStreamerTwitch;
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/misc/FollowHandler.java:55]:
```suggestion
        isStreamerFollowed = Service.isUserFollowingStreamer(mChannelInfo.getStreamerName(), mContext);
        return isStreamerFollowed;
    }
    public boolean isStreamerTwitch() {
        isStreamerTwitch = Service.isUserTwitch(mChannelInfo.getUserId(), mContext);
        return isStreamerTwitch;
```

File: app/src/main/java/com/perflyst/twire/misc/FollowHandler.java [hunk 3]
```diff
@@ -17,64 +14,53 @@
...
+        this.isStreamerTwitch = false;
 
         init();
     }
 
     private void init() {
         if (!new Settings(mContext).isLoggedIn()) {
-            mDelegate.userIsNotLoggedIn();
+            //
         } else {
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/misc/FollowHandler.java:33]: Why not invert the condition?

File: app/src/main/java/com/perflyst/twire/misc/FollowHandler.java [hunk 4]
```diff
@@ -17,64 +14,53 @@
...
             boolean isUserFollowed = Service.isUserFollowingStreamer(mChannelInfo.getStreamerName(), mContext);
+            boolean isUserTwitch = Service.isUserTwitch(mChannelInfo.getUserId(), mContext);
             if (isUserFollowed) {
                 isStreamerFollowed = true;
             }
+            if (isUserTwitch) {
+                isStreamerTwitch = true;
+            }
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/misc/FollowHandler.java:41]:
```suggestion
            if (isUserFollowed) {
                isStreamerFollowed = true;
            }
            isStreamerTwitch = Service.isUserTwitch(mChannelInfo.getUserId(), mContext);
```

File: app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java [hunk 5]
```diff
@@ -141,4 +144,60 @@ public void onClickFiltersStreamsByLanguageEnable(View v) {
...
+
+    // Export/Import for Follows
+
+    public void onExport(View v) {
+        MaterialDialog dialog = DialogService.getSettingsExportFollowsDialog(this);
+        dialog.getBuilder().onPositive((dialog1, which) -> {
+            dialog1.dismiss();
+            SubscriptionsDbHelper helper = new SubscriptionsDbHelper(getBaseContext());
+            int exported = helper.onExport(helper.getWritableDatabase());
+            Toast infoToast = Toast.makeText(getBaseContext(),"Exported " + exported + " Follows", Toast.LENGTH_SHORT);
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java:175]: This string is not localized.

File: app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java [hunk 6]
```diff
@@ -141,4 +144,60 @@ public void onClickFiltersStreamsByLanguageEnable(View v) {
...
+        dialog.show();
+    }
+
+    public void onImport(View v) {
+        MaterialDialog dialog = DialogService.getSettingsImportFollowsDialog(this);
+        dialog.getBuilder().onPositive((dialog1, which) -> {
+            dialog1.dismiss();
+            SubscriptionsDbHelper helper = new SubscriptionsDbHelper(getBaseContext());
+            int imported = helper.onImport(helper.getWritableDatabase());
+            Toast infoToast = Toast.makeText(getBaseContext(),"Imported " + imported + " Follows", Toast.LENGTH_SHORT);
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java:192]: This string is not localized.

File: app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java [hunk 7]
```diff
@@ -141,4 +144,60 @@ public void onClickFiltersStreamsByLanguageEnable(View v) {
...
+
+    // Database Stuff below
+
+    public void onClickWipeFollows(View v) {
+        MaterialDialog dialog = DialogService.getSettingsWipeFollowsDialog(this);
+        dialog.getBuilder().onPositive((dialog1, which) -> {
+            dialog1.dismiss();
+            SubscriptionsDbHelper helper = new SubscriptionsDbHelper(getBaseContext());
+            helper.onWipe(helper.getWritableDatabase(), settings.isLoggedIn());
+            Toast infoToast = Toast.makeText(getBaseContext(),"Wiped Follows", Toast.LENGTH_SHORT);
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java:156]: This string is not localized.

File: app/src/main/java/com/perflyst/twire/activities/ChannelActivity.java [hunk 8]
```diff
@@ -295,21 +295,29 @@ private void initiateFAB() {
...
 
-        mFab.setOnClickListener(v -> {
-            if (mFollowHandler.isStreamerFollowed()) {
-                mFollowHandler.unfollowStreamer();
-            } else {
-                mFollowHandler.followStreamer();
-            }
-
+        // If the channel got imported from Twitch, then hide the Follow/Unfollow Button
+        if (mFollowHandler.isStreamerFollowed() && mFollowHandler.isStreamerTwitch()) {
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/ChannelActivity.java:299]: Why does this check if the streamer is followed? Couldn't we just check if the follow is from Twitch?

File: app/src/main/java/com/perflyst/twire/activities/ChannelActivity.java [hunk 9]
```diff
@@ -295,21 +295,29 @@ private void initiateFAB() {
...
 
-        updateFABIcon(mFollowHandler.isStreamerFollowed());
+                hideFAB();
+                new Handler().postDelayed(() -> {
+                    updateFABIcon(mFollowHandler.isStreamerFollowed());
+                    if (mFollowHandler.isStreamerTwitch()) {
+                        hideFAB();
+                    } else {
+                        showFAB();
+                    }
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/ChannelActivity.java:316]: Why are we checking if the streamer is from Twitch? And if they are we're hiding the button, even though it's already hidden by line 309?

**Review by samfundev — CHANGES_REQUESTED** [review]
I haven't checked through everything, but I'm going to stop my review for now so that these initial comments can be worked on.

File: app/src/main/java/com/perflyst/twire/fragments/NavigationDrawerFragment.java [hunk 10]
```diff
@@ -245,11 +245,15 @@ private void checkUserLogin() {
...
 
+
+        /*
+        this was used to hide elements when the user is not logged in
         if (!mSettings.isLoggedIn()) {
             for (View userView : mUserRequiredViews) {
                 userView.setVisibility(View.GONE);
             }
         }
+         */
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/fragments/NavigationDrawerFragment.java]: Removed in e33e590f68a088e41086760ac775bbd4eb139984

File: app/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -202,6 +202,18 @@
...
+    <string name="gen_dialog_wipe_follows_cancel_action">Cancel</string>
+    <string name="gen_dialog_export_follows_content">Do you want to Export all Follows?</string>
+    <string name="gen_dialog_export_follows_action">Export</string>
+    <string name="gen_dialog_export_follows_cancel_action">Cancel</string>
+    <string name="gen_dialog_import_follows_content">Do you want to Import all Follows?</string>
+    <string name="gen_dialog_import_follows_action">Import</string>
+    <string name="gen_dialog_import_follows_cancel_action">Cancel</string>
+    <string name="gen_twitch_follows">Wipe all Follows</string>
+    <string name="gen_export_database">Export all Follows</string>
+    <string name="gen_import_database">Import Follows</string>
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/res/values/strings.xml]: Changed in f45b1f1c7f2c0da7213d08525d1366fa8d00bc4a

File: app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java (hunk 5 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java]: Changed in 16694b36622fa1e6aaf50eff17ee1ea7c8d6c090

File: app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java (hunk 7 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java]: Changed in 16694b36622fa1e6aaf50eff17ee1ea7c8d6c090

File: app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java (hunk 6 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java]: Changed in 16694b36622fa1e6aaf50eff17ee1ea7c8d6c090

File: app/src/main/java/com/perflyst/twire/activities/ChannelActivity.java (hunk 8 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/ChannelActivity.java]: Changed in df0f06c1b0229e65d28da947cc3db8443033bfee

File: app/src/main/java/com/perflyst/twire/activities/ChannelActivity.java (hunk 9 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/ChannelActivity.java]: Duplicate check removed in df0f06c1b0229e65d28da947cc3db8443033bfee

File: app/src/main/java/com/perflyst/twire/misc/FollowHandler.java (hunk 2 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/misc/FollowHandler.java]: Changed in 96e422e91645f6ec8afb3565efd90a543bc9e742

File: app/src/main/java/com/perflyst/twire/misc/FollowHandler.java (hunk 4 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/misc/FollowHandler.java]: Changed in 96e422e91645f6ec8afb3565efd90a543bc9e742

File: app/src/main/java/com/perflyst/twire/misc/FollowHandler.java (hunk 3 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/misc/FollowHandler.java]: Inverted in 96e422e91645f6ec8afb3565efd90a543bc9e742

File: app/src/main/java/com/perflyst/twire/tasks/GetStreamsCountTask.java (hunk 1 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/tasks/GetStreamsCountTask.java]: I have changed the API to Helix and support for more than 100 Channel in 3a1684851a9ccf94757efdb861886fb0d354fe9e

File: app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java [hunk 12]
```diff
@@ -51,12 +54,25 @@ public void addToAdapter(List<StreamInfo> aObjectList) {
     }
 
     @Override
-    public List<StreamInfo> getVisualElements() throws JSONException, MalformedURLException {
-        final String URL = "https://api.twitch.tv/kraken/streams/followed?oauth_token=" + new Settings(getBaseContext()).getGeneralTwitchAccessToken() + "&limit=" + getLimit() + "&offset=" + getCurrentOffset() + "&stream_type=live";
+    public List<StreamInfo> getVisualElements() throws JSONException, ExecutionException, InterruptedException, MalformedURLException {
+        // build the api link
+        String kraken_url = "https://api.twitch.tv/kraken/streams?limit=" + getLimit() + "&stream_type=live&offset=" + getCurrentOffset()+ "&channel=";
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java:59]: This should also be moved to the helix API.

File: app/src/main/java/com/perflyst/twire/service/SubscriptionsDbHelper.java [hunk 13]
```diff
@@ -87,4 +106,187 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
...
+        cursor.close();
+
+        if (channelstoExport.length()>0) {
+            try {
+                JSONObject channels = new JSONObject();
+                channels.put("Channels", channelstoExport);
+                String jsonStr = channels.toString();
+                Log.d("Export String", jsonStr);
+
+                // TO-DO Save File
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/service/SubscriptionsDbHelper.java:239]: This TODO seems like it's completed.

**Review by samfundev — CHANGES_REQUESTED** [review]
Noticed a few more things.

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -206,6 +206,19 @@
     <string name="gen_tips_have_been_reset">Tutorial tips have been reset</string>
     <string name="gen_language_filter">Filter top streams by language</string>
     <string name="gen_open_changelog">Open changelog</string>
+    <string name="gen_dialog_wipe_follows_content">Do you want to wipe all follows?</string>
+    <string name="gen_dialog_wipe_follows_action">Wipe</string>
+    <string name="gen_dialog_follows_cancel_action">Cancel</string>
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/res/values/strings.xml]: Changed in c054689cca0ff21294051ecd00c5ef5eb52586d6

File: app/src/main/res/layout/activity_settings_general.xml [hunk 15]
```diff
@@ -273,6 +273,97 @@
...
+            <View
+                android:layout_width="match_parent"
+                android:layout_height="@dimen/settings_divider_height"
+                android:background="?attr/dividerColor" />
+				
+				
+			<View
+                android:layout_width="match_parent"
+                android:layout_height="@dimen/settings_divider_height"
+                android:background="?attr/dividerColor" />
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/res/layout/activity_settings_general.xml]: No, removed in f0cedc3d0803a96a154a4a8f57ddf29599dc1ba4

File: app/src/main/java/com/perflyst/twire/service/SubscriptionsDbHelper.java (hunk 13 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/service/SubscriptionsDbHelper.java]: Removed in eba321f25a61ac172f27085cbbb03c8bc78f6632

File: app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java (hunk 12 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java]: Helix is added in 8bef94a0772d842ff4c2a265077a7185794122a6, 995bb9dd172294a895fefa34fd02f4f5f6ebba23 and ce98da971618bc54a322f84b516ba9f66dbb337b

**Comment by samfundev** [comment]
@Nickwasused Should I continue reviewing this or are you still working on it? Some stuff is still unresolved from the previous review.

File: app/src/main/java/com/perflyst/twire/activities/main/MyChannelsActivity.java [hunk 16]
```diff
@@ -48,9 +48,12 @@ public void addToAdapter(List<ChannelInfo> aObjectList) {
 
     @Override
     public List<ChannelInfo> getVisualElements() throws ExecutionException, InterruptedException {
+        /*
+        This is not updating, for the moment we are going to force reload it every time
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/main/MyChannelsActivity.java]: Well I don´t exactly remember what i meant with that, but it got reverted in 36fcec83311ff556e1c8bf578622271cfe6e4c22

File: app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java [hunk 17]
```diff
@@ -169,6 +172,61 @@ public void onClickOpenChangelog(View v) {
...
+            int imported = helper.onImport(helper.getWritableDatabase());
+            Toast infoToast = Toast.makeText(getBaseContext(), String.format(getString(R.string.gen_toast_import_database), imported), Toast.LENGTH_SHORT);
+            infoToast.show();
+        });
+
+        dialog.getBuilder().onNegative((dialog2, which) -> {
+            dialog2.dismiss();
+        });
+
+        dialog.show();
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/settings/SettingsGeneralActivity.java]: Changed in 0729d3fbffc0ba8ed77461fe070e005d9f521652

**Comment by Nickwasused** [comment]
@samfundev I think you can start to review, the only thing left is to move the search to helix.

File: app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java [hunk 18]
```diff
@@ -51,17 +55,64 @@ public void addToAdapter(List<StreamInfo> aObjectList) {
...
+            JSONObject fullDataObject = new JSONObject(temp_jsonString);
+            // create the array
+            JSONArray temp_array = fullDataObject.getJSONArray(ARRAY_KEY);
+            // append the new array to the final one
+            for (int x=0; x<temp_array.length(); x++) {
+                final_array.put(temp_array.get(x));
+            }
+        }
 
         List<StreamInfo> mResultList = new ArrayList<>();
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java:112]: Guava has a really nice method that makes this much simplier.
```suggestion
        String helix_url = "https://api.twitch.tv/helix/streams?first=100";

        ArrayList<ChannelInfo> channels;
        if (TempStorage.hasLoadedStreamers()) {
            channels = new ArrayList<>(TempStorage.getLoadedStreamers());
        } else {
            GetFollowsFromDB subscriptionsTask = new GetFollowsFromDB();
            subscriptionsTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, getBaseContext());
            channels = new ArrayList<>(subscriptionsTask.get().values());
        }

        // loop over all channel in the DB in chunks of 100
        List<StreamInfo> mResultList = new ArrayList<>();
        final String ARRAY_KEY = "data";
        for (List<ChannelInfo> chunk : Lists.partition(channels, 100)) {
            String url = helix_url + Lists.transform(chunk, channelInfo -> "&user_id=" + channelInfo.getUserId());

            // request the url
            String temp_jsonString = Service.urlToJSONStringHelix(url, this);
            JSONObject fullDataObject = new JSONObject(temp_jsonString);
            // create the array
            JSONArray temp_array = fullDataObject.getJSONArray(ARRAY_KEY);
            // append the new array to the final one
            for (int i = 0; i < temp_array.length(); i++) {
                JSONObject streamObject = temp_array.getJSONObject(i);
                mResultList.add(JSONService.getStreamInfo(getBaseContext(), streamObject, null, false));
            }
        }
```

File: app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java [hunk 19]
```diff
@@ -51,17 +55,64 @@ public void addToAdapter(List<StreamInfo> aObjectList) {
...
-        String jsonString = Service.urlToJSONString(URL);
-        JSONObject fullDataObject = new JSONObject(jsonString);
-        JSONArray topStreamsArray = fullDataObject.getJSONArray(ARRAY_KEY);
 
-        for (int i = 0; i < topStreamsArray.length(); i++) {
-            JSONObject streamObject = topStreamsArray.getJSONObject(i);
+        for (int i = 0; i < final_array.length(); i++) {
+            JSONObject streamObject = final_array.getJSONObject(i);
             mResultList.add(JSONService.getStreamInfo(getBaseContext(), streamObject, null, false));
         }
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java:118]: This is covered by the other suggestion and can be deleted. (For some reason, I can't just make this one suggestion)
```suggestion
```

**Review by samfundev — CHANGES_REQUESTED** [review]
Hopefully the last review.

File: app/src/main/java/com/perflyst/twire/service/JSONService.java [hunk 20]
```diff
@@ -11,80 +11,115 @@
...
+import java.util.regex.Pattern;
 
 /**
  * Created by Sebastian Rask on 25-04-2016.
  */
 public class JSONService {
     private static final String LOG_TAG = JSONService.class.getSimpleName();
 
+    // based on https://github.com/twurple/twurple/blob/6d3ca508fe0a21fadd77b63b62d0df66b9150f97/packages/api/src/api/helix/video/HelixVideo.ts#L175
+    public static int getVodlenght(String length) {
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/service/JSONService.java]: Changed in 43230f31ada33ab919dcfc9d5634d215e85488c5

File: app/src/main/java/com/perflyst/twire/activities/main/MyChannelsActivity.java [hunk 21]
```diff
@@ -50,11 +50,10 @@ public void addToAdapter(List<ChannelInfo> aObjectList) {
     public List<ChannelInfo> getVisualElements() throws ExecutionException, InterruptedException {
         if (TempStorage.hasLoadedStreamers()) {
             return new ArrayList<>(TempStorage.getLoadedStreamers());
+        } else {
+            GetFollowsFromDB subscriptionsTask = new GetFollowsFromDB();
+            subscriptionsTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, getBaseContext());
+            return new ArrayList<>(subscriptionsTask.get().values());
```
[the remark this answers is not in the document]
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/main/MyChannelsActivity.java]: Changed back in 2b537088036a1eccbf901b5fc0684c5470854bf8

File: app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java (hunk 19 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java]: Done in e7dc3eddc07ef5f0e0b5710432a050e6010be40e

File: app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java [hunk 22]
```diff
@@ -51,18 +56,41 @@ public void addToAdapter(List<StreamInfo> aObjectList) {
...
+        final String ARRAY_KEY = "data";
+        for (List<ChannelInfo> chunk : Lists.partition(channels, 100)) {
+            // this returns a list, I found no easy way to convert it to a string
+            List ids = Lists.transform(chunk, channelInfo -> "&user_id=" + channelInfo.getUserId());
+            // convert the new list to a string
+            String id_string = "";
+            for (int i=0; i<ids.size(); i++) {
+                id_string = id_string + ids.get(i);
+            }
+            String url = helix_url + id_string;
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java:83]: Oh sorry about that, there is a easy way to make the list into a string. I forgot to use that in my suggestion.
```suggestion
            String url = helix_url + Joiner.on("").join(Lists.transform(chunk, channelInfo -> "&user_id=" + channelInfo.getUserId()));
```

Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/activities/main/MyStreamsActivity.java]: Thanks! Changed in c500ea5546059102be290b92bcced70a51ca3a90

**Comment by samfundev** [comment]
@Nickwasused Oh wait, are you going to migrate the search to helix as well?

**Comment by Nickwasused** [comment]
@samfundev It would be great if you can take a look at it.

**Pull request merged**